aptos-move/aptos-aggregator/src/delta_change_set.rs

```diff
@@ -64,54 +51,53 @@ impl DeltaOp {
         // this is possible if the values observed during execution didn't
         // overflow or dropped below zero. The check can be emulated by actually
         // doing addition and subtraction.
-        addition(base, self.max_positive, self.limit)?;
-        subtraction(base, self.min_negative)?;
+        addition(
+            base,
+            self.history.max_achieved_positive_delta,
+            self.max_value,
+        )?;
+        subtraction(base, self.history.min_achieved_negative_delta)?;
+
+        // Do we need to check overflow/underflow from history as well?
 
         // If delta has been successfully validated, apply the update.
-        match self.update {
-            DeltaUpdate::Plus(value) => addition(base, value, self.limit),
-            DeltaUpdate::Minus(value) => subtraction(base, value),
-        }
+        Ok(addition_deltavalue(base, self.update, self.max_value)?)
     }
 
     /// Shifts by a `delta` the maximum positive value seen by `self`.
-    fn shifted_max_positive_by(&self, delta: &DeltaOp) -> PartialVMResult<u128> {
-        match delta.update {
-            // Suppose that maximum value seen is +M and we shift by +V. Then the
-            // new maximum value is M+V provided addition do no overflow.
-            DeltaUpdate::Plus(value) => addition(value, self.max_positive, self.limit),
-            // Suppose that maximum value seen is +M and we shift by -V this time.
-            // If M >= V, the result is +(M-V). Otherwise, `self` should have never
-            // reached any positive value. By convention, we use 0 for the latter
-            // case. Also, we can reuse `subtraction` which throws an error when M < V,
-            // simply mapping the error to 0.
-            DeltaUpdate::Minus(value) => Ok(subtraction(self.max_positive, value).unwrap_or(0)),
-        }
+    fn shifted_max_achieved_positive_delta_by(&self, delta: &DeltaOp) -> PartialVMResult<u128> {
```

Comment:
Let's re-use the code for all this and below (shifting, merging) things from AggregatorChange merging?(merge_with_previous_aggregator_change), I added a comment there as well.
@igor-aptos might be related to the discussion of what types to keep (e.g. if we unset speculative value, maybe we can get rid of DeltaOp altogether - at least for V2? - may end up ugly though to have an enum of AggregatorChange or old DeltaOp, so I'm actually fine as is as long as we share the code)